src/payout_tokens.cc

```diff
@@ -0,0 +1,183 @@
+
+    auto verification_signature = verification_key.sign(payload_json);
+    auto verification_signature_base64 = verification_signature.encode_base64();
+    credential_dictionary.SetKey("signature",
+        base::Value(verification_signature_base64));
+
+    auto preimage = unblinded_payment_token.preimage();
+    auto preimage_base64 = preimage.encode_base64();
+    credential_dictionary.SetKey("t", base::Value(preimage_base64));
+
+    // Create payment credential
+    auto* payment_credential_dictionary = new base::DictionaryValue();
+
+    payment_credential_dictionary->SetKey("credential",
+        base::DictionaryValue(std::move(credential_dictionary)));
+
+    auto public_key_base64 = public_key_.encode_base64();
+    payment_credential_dictionary->SetKey("publicKey",
+        base::Value(public_key_base64));
+
+    payment_credentials_list->Append(
+        std::unique_ptr<base::DictionaryValue>(payment_credential_dictionary));
+  }
+
+  // Create request body
+  base::DictionaryValue body_dictionary;
+
+  body_dictionary.SetWithoutPathExpansion("paymentCredentials",
+      std::move(payment_credentials_list));
+
+  body_dictionary.SetKey("payload", base::Value(payload_json));
+
+  std::string json;
+  base::JSONWriter::Write(body_dictionary, &json);
+  std::string body = json;
+
+  // Create request headers
+  std::vector<std::string> headers = {};
+
+  auto accept_header = std::string("accept: ").append("application/json");
+  headers.push_back(accept_header);
+
+  // Create request content type
+  std::string content_type = "application/json";
+
+  // PUT /v1/confirmation/payment/{payment_id}
+  BLOG(INFO) << "PUT /v1/confirmation/payment/{payment_id}";
+
+  auto endpoint = std::string("/v1/confirmation/payment/").append(payment_id_);
+  auto ads_serve_url = helper::AdsServe::GetURL().append(endpoint);
+
+  BLOG(INFO) << "URL Request:";
+  BLOG(INFO) << "  URL: " << ads_serve_url;
+  BLOG(INFO) << "  Headers:";
+  for (const auto& header : headers) {
+    BLOG(INFO) << "    " << header;
+  }
+  BLOG(INFO) << "  Body: " << body;
+  BLOG(INFO) << "  Content_type: " << content_type;
+
+  auto callback = std::bind(&PayoutTokens::OnRedeemPaymentTokens,
+      this, ads_serve_url, _1, _2, _3);
+
+  confirmations_client_->URLRequest(ads_serve_url, headers, body, content_type,
+      URLRequestMethod::PUT, callback);
+}
+
+void PayoutTokens::OnRedeemPaymentTokens(
+    const std::string& url,
+    const int response_status_code,
+    const std::string& response,
+    const std::map<std::string, std::string>& headers) {
+  BLOG(INFO) << "OnRedeemPaymentTokens";
+
+  BLOG(INFO) << "URL Request Response:";
+  BLOG(INFO) << "  URL: " << url;
+  BLOG(INFO) << "  Response Status Code: " << response_status_code;
+  BLOG(INFO) << "  Response: " << response;
+  BLOG(INFO) << "  Headers:";
+  for (const auto& header : headers) {
+    BLOG(INFO) << "    " << header.first << ": " << header.second;
+  }
+
+  if (response_status_code != 200) {
```

Comment:
Are any other 2XX response codes possible here?

Reply:
Not according to original code, and checking with sever team on a previous occasion

Reply:
I have checked and we are covering all response status codes